src/Microsoft.AspNetCore.Razor.Evolution/DefaultRazorIRLoweringPhase.cs

```diff
@@ -399,27 +404,41 @@ public override void VisitMarkupSpan(MarkupChunkGenerator chunkGenerator, Span s
                 if (currentChildren.Count > 0 && currentChildren[currentChildren.Count - 1] is HtmlContentIRNode)
                 {
                     var existingHtmlContent = (HtmlContentIRNode)currentChildren[currentChildren.Count - 1];
-                    existingHtmlContent.Content = string.Concat(existingHtmlContent.Content, span.Content);
 
-                    if (existingHtmlContent.Source != null)
+                    var source = BuildSourceSpanFromNode(span);
+                    if (existingHtmlContent.Source == null && source == null)
                     {
-                        var contentLength = existingHtmlContent.Source.Value.Length + span.Content.Length;
+                        Combine(existingHtmlContent, span);
+                        return;
+                    }
 
-                        existingHtmlContent.Source = new SourceSpan(
-                            existingHtmlContent.Source.Value.FilePath ?? Filename,
-                            existingHtmlContent.Source.Value.AbsoluteIndex,
-                            existingHtmlContent.Source.Value.LineIndex,
-                            existingHtmlContent.Source.Value.CharacterIndex,
-                            contentLength);
+                    if (source != null &&
+                        existingHtmlContent.Source != null &&
+                        existingHtmlContent.Source.Value.FilePath == source.Value.FilePath &&
+                        existingHtmlContent.Source.Value.AbsoluteIndex + existingHtmlContent.Source.Value.Length == source.Value.AbsoluteIndex)
+                    {
+                        Combine(existingHtmlContent, span);
+                        return;
                     }
                 }
-                else
+
+                _builder.Add(new HtmlContentIRNode()
                 {
-                    _builder.Add(new HtmlContentIRNode()
-                    {
-                        Content = span.Content,
-                        Source = BuildSourceSpanFromNode(span),
-                    });
+                    Content = span.Content,
+                    Source = BuildSourceSpanFromNode(span),
+                });
+            }
+            private void Combine(HtmlContentIRNode node, Span span)
+            {
+                node.Content = node.Content + span.Content;
+                if (node.Source != null)
+                {
+                    node.Source = new SourceSpan(
+                        node.Source.Value.FilePath ?? Filename,
```

Comment:
So we don't have this magic in two places we could re-purpose BuildSourceSpanFromNode to have an alternative that takes in a SourceSpan and a length.

It'd be nice to ensure there's only 1 location doing this file path logic is all.